## ✨ 크롤링 데이터 ✨

### 1. 데이터 EDA

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("/Users/jeonhaelee/Documents/boostcamp/project/code/Crawling/Personality_All_168000_rows.csv")
df.head()

Character                Contents      Votes  MBTI Enneagram  \
0              Makima            Chainsaw Man   16.1k/2k  ENFJ       2w1   
1  Daenerys Targaryen  Game of Thrones (2011)   5.9k/940  ENFJ       1w2   
2      Tanjiro Kamado        Kimetsu no Yaiba   5.7k/799  ENFJ       1w2   
3        Oikawa Tooru               Haikyuu!!   5.6k/683  ENFJ       3w4   
4          Homelander    The Boys (2019-2022)  5.2k/1.9k  ENFJ       3w4   

                                            category  \
0  https://pdb-web-static.s3.us-east-2.amazonaws....   
1  https://pdb-web-static.s3.us-east-2.amazonaws....   
2  https://pdb-web-static.s3.us-east-2.amazonaws....   
3  https://pdb-web-static.s3.us-east-2.amazonaws....   
4  https://pdb-web-static.s3.us-east-2.amazonaws....   

                                             img_src  Page  
0  https://static1.personality-database.com/profi...     1  
1  https://static1.personality-database.com/profi...     1  
2  https://static1.personality-database.com/profi...     1  
3  https://static1.personality-database.com/profi...     1  
4  https://static1.personality-database.com/profi...     1

- 캐릭터명이 중복된 행은 총 71827의 행
- 27220개의 고유 캐릭터가 중복되어있음


In [211]:
df

Character                  Contents  \
0                                          Makima              Chainsaw Man   
1                              Daenerys Targaryen    Game of Thrones (2011)   
2                                  Tanjiro Kamado          Kimetsu no Yaiba   
3                                    Oikawa Tooru                 Haikyuu!!   
4                                      Homelander      The Boys (2019-2022)   
...                                           ...                       ...   
167995                            Amelia Scanwell                   Harlots   
167996                                        Jim              Toaster Dude   
167997                                  Channeler      Pokémon Red and Blue   
167998                                     Raksha           The Jungle Book   
167999  Abdul Qadir Al Jilani, Hanbali Theologian  Islamic Scholars (Sunni)   

            Votes  MBTI Enneagram       category  \
0        16.1k/2k  ENFJ       2w1  anime & manga   
1        5.9k/940  ENFJ       1w2     television   
2        5.7k/799  ENFJ       1w2  anime & manga   
3        5.6k/683  ENFJ       3w4  anime & manga   
4       5.2k/1.9k  ENFJ       3w4     television   
...           ...   ...       ...            ...   
167995        6/0  INFJ       9w1     television   
167996        6/0  INFJ       5w4      web comic   
167997        6/2  INFJ       2w1              n   
167998        6/0  INFJ       XwX              n   
167999        6/0  INFJ       9w1              n   

                                                  img_src  Page  
0       https://static1.personality-database.com/profi...     1  
1       https://static1.personality-database.com/profi...     1  
2       https://static1.personality-database.com/profi...     1  
3       https://static1.personality-database.com/profi...     1  
4       https://static1.personality-database.com/profi...     1  
...                                                   ...   ...  
167995  https://pdb-web-static.s3.us-east-2.amazonaws....   210  
167996  https://static1.personality-database.com/profi...   210  
167997  https://static1.personality-database.com/profi...   210  
167998  https://static1.personality-database.com/profi...   210  
167999  https://static1.personality-database.com/profi...   210  

[168000 rows x 8 columns]

In [213]:
df[df['category'] == 'movie']['Contents'].nunique()

4291

In [132]:
df[df.duplicated(subset='Character', keep=False)]

Character                                 Contents  \
1       Daenerys Targaryen                   Game of Thrones (2011)   
2           Tanjiro Kamado                         Kimetsu no Yaiba   
4               Homelander                     The Boys (2019-2022)   
5                   Monika                Doki Doki Literature Club   
7                     Emma                    Yakusoku no Neverland   
...                    ...                                      ...   
167972            The Girl  A Girl Walks Home Alone at Night (2014)   
167978             Quaithe          A Song of Ice and Fire (Series)   
167989              Xandra                                  Neopets   
167996                 Jim                             Toaster Dude   
167998              Raksha                          The Jungle Book   

            Votes  MBTI Enneagram       category  \
1        5.9k/940  ENFJ       1w2     television   
2        5.7k/799  ENFJ       1w2  anime & manga   
4       5.2k/1.9k  ENFJ       3w4     television   
5        4.7k/626  ENFJ       2w3              n   
7        3.9k/917  ENFJ       2w1  anime & manga   
...           ...   ...       ...            ...   
167972        6/0  INFJ       4w5          movie   
167978        6/0  INFJ       6w5              n   
167989        6/0  INFJ       6w5              n   
167996        6/0  INFJ       5w4      web comic   
167998        6/0  INFJ       XwX              n   

                                                  img_src  Page  
1       https://static1.personality-database.com/profi...     1  
2       https://static1.personality-database.com/profi...     1  
4       https://static1.personality-database.com/profi...     1  
5       https://static1.personality-database.com/profi...     1  
7       https://static1.personality-database.com/profi...     1  
...                                                   ...   ...  
167972  https://static1.personality-database.com/profi...   210  
167978  https://static1.personality-database.com/profi...   210  
167989  https://static1.personality-database.com/profi...   210  
167996  https://static1.personality-database.com/profi...   210  
167998  https://static1.personality-database.com/profi...   210  

[71827 rows x 8 columns]

In [134]:
df[df.duplicated(subset='Character', keep=False)]['Character'].nunique()

27220

In [133]:
df[df.duplicated(subset='Character', keep=False)]['Character'].value_counts()

Sam                    67
Alice                  59
Lucy                   55
Alex                   54
Jack                   51
                       ..
Boss Mo                 2
Kazuki Fuuchouin        2
Sejun (VICTON)          2
Telolilo Lilili Lu      2
Srinivasa Ramanujan     2
Name: Character, Length: 27220, dtype: int64

- 같은 이름을 가진 캐릭터가 여러개임을 확인.
- 컨텐츠 명까지 조건 추가해서 중복확인 다시 해주자.

In [135]:
df[df.duplicated(subset=['Character', 'Contents'], keep=False)]

Character                                 Contents     Votes  \
71          Rachel Amber                          Life is Strange  1.1k/218   
173       Shion Sonozaki                Higurashi no Naku Koro ni    510/61   
199      Sunny Starscout  My Little Pony: A New Generation (2021)    459/83   
200      Sunny Starscout  My Little Pony: A New Generation (2021)    459/83   
298              Gabriel                    The Mandela Catalogue    324/94   
...                  ...                                      ...       ...   
167881  Remedios Infante                        La Que Se Avecina       7/0   
167882              Oğuz                Gelsin Hayat Bildiği Gibi       7/0   
167883              Gray                               Blood Kiss       7/0   
167884             Ou Li                    Falling Into Darkness       7/0   
167885     Mario Mendoza             Writers (Literature, Modern)       7/0   

        MBTI Enneagram    category  \
71      ENFJ       4w3           n   
173     ENFJ       4w5           n   
199     ENFJ       1w2       movie   
200     ENFJ       1w2       movie   
298     ENFJ       8w7     cartoon   
...      ...       ...         ...   
167881  INFJ       XwX  television   
167882  INFJ       1w9  television   
167883  INFJ       4w5           n   
167884  INFJ       6w5   web comic   
167885  INFJ       XwX           n   

                                                  img_src  Page  
71      https://static1.personality-database.com/profi...     2  
173     https://static1.personality-database.com/profi...     4  
199     https://static1.personality-database.com/profi...     4  
200     https://static1.personality-database.com/profi...     5  
298     https://static1.personality-database.com/profi...     6  
...                                                   ...   ...  
167881  https://static1.personality-database.com/profi...   208  
167882  https://pdb-web-static.s3.us-east-2.amazonaws....   208  
167883  https://pdb-web-static.s3.us-east-2.amazonaws....   208  
167884  https://static1.personality-database.com/profi...   208  
167885  https://pdb-web-static.s3.us-east-2.amazonaws....   208  

[48295 rows x 8 columns]

- 중복 행이 48295 행으로 줄었다!
- 22236의 고유 캐릭터가 중복되어 있음

In [136]:
df[df.duplicated(subset=['Character', 'Contents'], keep=False)]['Character'].nunique()

22236

In [137]:
df[df.duplicated(subset=['Character', 'Contents'], keep=False)]['Character'].value_counts()

Sam                 22
Olivia              21
Lucy                20
Mary                20
Tom                 20
                    ..
Zeta                 2
Clarso               2
Liberty Prime        2
Donald Davenport     2
Madoka Kabuya        2
Name: Character, Length: 22236, dtype: int64

In [138]:
df[df.duplicated(subset=['Character', 'Contents'], keep=False)].sort_values(by='Character')

Character  \
59367            "Arthur Slugworth"   
59541            "Arthur Slugworth"   
82023                        "Beta"   
82432                        "Beta"   
3534   "Black Sam" Bellamy (Pirate)   
...                             ...   
51794                 黄蓉 Huang Rong   
83679                           민수호   
83882                           민수호   
93708                           조예지   
93645                           조예지   

                                           Contents Votes  MBTI Enneagram  \
59367  Willy Wonka and the Chocolate Factory (1971)  12/3  INTJ       4w5   
59541  Willy Wonka and the Chocolate Factory (1971)  12/3  INTJ       4w5   
82023                       The Walking Dead Comics   8/0  ISTJ       1w9   
82432                       The Walking Dead Comics   8/0  ISTJ       1w9   
3534                     Historical Figures (1700s)  24/1  ENFJ       6w7   
...                                             ...   ...   ...       ...   
51794         射雕英雄传 The Legend of the Condor Heroes  10/0  ENTP       7w6   
83679                                       Closers   6/0  ISTJ       6w5   
83882                                       Closers   6/0  ISTJ       6w5   
93708                             Zombie Highschool   6/0  ESTJ       3w4   
93645                             Zombie Highschool   6/0  ESTJ       3w4   

      category                                            img_src  Page  
59367    movie  https://static1.personality-database.com/profi...   138  
59541    movie  https://static1.personality-database.com/profi...   141  
82023    comic  https://static1.personality-database.com/profi...   171  
82432    comic  https://static1.personality-database.com/profi...   179  
3534         n  https://static1.personality-database.com/profi...    71  
...        ...                                                ...   ...  
51794        n  https://static1.personality-database.com/profi...   196  
83679        n  https://pdb-web-static.s3.us-east-2.amazonaws....   204  
83882        n  https://pdb-web-static.s3.us-east-2.amazonaws....   208  
93708        n  https://static1.personality-database.com/profi...   195  
93645        n  https://static1.personality-database.com/profi...   193  

[48295 rows x 8 columns]

- 대부분의 중복이 페이지 수가 달라서 임을 확인
- 캐릭터 명도 클렌징 필요 -> " " 처리, 언어 처리 등등...
- 애니어그램 클렌징 필요

#### Lucy 캐릭터로 파악

In [139]:
dupli_character = df[df.duplicated(subset=['Character', 'Contents'], keep=False)]

In [140]:
dupli_character[dupli_character['Character'] == 'Lucy']

Character                                Contents Votes  MBTI  \
37589       Lucy                      Summer Camp Island  14/1  INTP   
37615       Lucy                      Summer Camp Island  14/1  INTP   
90885       Lucy                         Bird Box (2018)  10/0  ESTJ   
90908       Lucy                         Bird Box (2018)  10/0  ESTJ   
112107      Lucy            Sense and Sensibility (1995)   9/2  ESFJ   
112165      Lucy            Sense and Sensibility (1995)   9/2  ESFJ   
127892      Lucy                          Subway Surfers  59/4  ISFP   
127908      Lucy                          Subway Surfers  59/4  ISFP   
133809      Lucy            The Inbetweeners (UK) (2008)  10/2  ISFP   
134012      Lucy            The Inbetweeners (UK) (2008)  10/2  ISFP   
134848      Lucy  Lucy: The Daughter of the Devil (2005)   8/0  ISFP   
135047      Lucy  Lucy: The Daughter of the Devil (2005)   8/0  ISFP   
135695      Lucy                                 In Fear   7/0  ISFP   
136105      Lucy                                 In Fear   7/0  ISFP   
152779      Lucy                     Love, Victor (2020)  15/0  ISTP   
152800      Lucy                     Love, Victor (2020)  15/0  ISTP   
152855      Lucy                     Love, Victor (2020)  15/0  ISTP   
167362      Lucy                             Lion (2016)   7/0  INFJ   
167412      Lucy                             Lion (2016)   7/0  INFJ   
167571      Lucy                             Lion (2016)   7/0  INFJ   

       Enneagram    category  \
37589        5w4     cartoon   
37615        5w4     cartoon   
90885        XwX       movie   
90908        XwX       movie   
112107       3w2       movie   
112165       3w2       movie   
127892       4w3           n   
127908       4w3           n   
133809       6w5  television   
134012       6w5  television   
134848       XwX     cartoon   
135047       XwX     cartoon   
135695       4w5       movie   
136105       4w5       movie   
152779       6w5  television   
152800       6w5  television   
152855       6w5  television   
167362       5w6       movie   
167412       5w6       movie   
167571       5w6       movie   

                                                  img_src  Page  
37589   https://static1.personality-database.com/profi...   122  
37615   https://static1.personality-database.com/profi...   123  
90885   https://static1.personality-database.com/profi...   138  
90908   https://static1.personality-database.com/profi...   139  
112107  https://static1.personality-database.com/profi...   143  
112165  https://static1.personality-database.com/profi...   144  
127892  https://static1.personality-database.com/profi...    38  
127908  https://static1.personality-database.com/profi...    39  
133809  https://static1.personality-database.com/profi...   157  
134012  https://static1.personality-database.com/profi...   161  
134848  https://static1.personality-database.com/profi...   177  
135047  https://static1.personality-database.com/profi...   181  
135695  https://static1.personality-database.com/profi...   194  
136105  https://static1.personality-database.com/profi...   203  
152779  https://static1.personality-database.com/profi...   116  
152800  https://static1.personality-database.com/profi...   117  
152855  https://static1.personality-database.com/profi...   118  
167362  https://static1.personality-database.com/profi...   198  
167412  https://static1.personality-database.com/profi...   199  
167571  https://static1.personality-database.com/profi...   202

In [141]:
dupli_character[dupli_character['Character'] == 'Lucy'].shape[0]

20

In [144]:
dupli_character[dupli_character['Character'] == 'Lucy']['Contents'].value_counts()

Love, Victor (2020)                       3
Lion (2016)                               3
Bird Box (2018)                           2
Summer Camp Island                        2
Sense and Sensibility (1995)              2
Subway Surfers                            2
Lucy: The Daughter of the Devil (2005)    2
In Fear                                   2
The Inbetweeners (UK) (2008)              2
Name: Contents, dtype: int64

#### 카테고리 처리

In [62]:
df['category'].unique()

array(['https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_8.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_2.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_11.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_7.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_29.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_14.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_3.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_1.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_27.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_15.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_13.png',
       'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_ico

In [3]:
category_dic = {
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_8.png' : 'anime & manga', # 애니 (예 : 하이큐 ...)
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_2.png' : 'television', # 드라마, 예능 
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_11.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_7.png' : 'cartoon', # 카툰 (티비 애니)
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_29.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_14.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_3.png' : 'movie', # 영화
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_1.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_27.png' : 'n', # <- 확인 필요
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_15.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_13.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_12.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_26.png' : 'web comic', # 웹툰 (예 : 네이버 웹툰, 카카오 웹툰 ...)
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_16.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_17.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_22.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_18.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_9.png' : 'comic', # 만화책
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_5.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_21.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_28.png' : 'n',
    'https://pdb-web-static.s3.us-east-2.amazonaws.com/category_icon/cat_10.png' : 'n'
}

In [67]:
dupli_character['category'] = dupli_character['category'].map(category_dic)

/var/folders/r7/pyt9qwrd5t13xvh1tc1pyhfh0000gn/T/ipykernel_77069/1980566896.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dupli_character['category'] = dupli_character['category'].map(category_dic)


In [68]:
dupli_character.head()

Character                                 Contents   Votes  MBTI  \
200  Sunny Starscout  My Little Pony: A New Generation (2021)  459/83  ENFJ   
300          Gabriel                    The Mandela Catalogue  324/94  ENFJ   
400  Haruka Sawamura                                   Yakuza   247/8  ENFJ   
501  Nibutani Shinka           Chuunibyou demo Koi ga Shitai!   201/6  ENFJ   
597   Shion Sonozaki                Higurashi no Naku Koro ni  173/26  ENFJ   

    Enneagram       category  \
200       1w2          movie   
300       8w7        cartoon   
400       2w3              n   
501       3w2  anime & manga   
597       4w5  anime & manga   

                                               img_src  Page  
200  https://static1.personality-database.com/profi...     5  
300  https://static1.personality-database.com/profi...     7  
400  https://static1.personality-database.com/profi...     9  
501  https://static1.personality-database.com/profi...    11  
597  https://static1.personality-database.com/profi...    12

In [145]:
dupli_movie_df = dupli_character[dupli_character['category'] == 'movie']

In [146]:
dupli_movie_df # category가 movie인 데이터 중 Character, Contents가 중복된 데이터들

Character                                 Contents   Votes  \
199     Sunny Starscout  My Little Pony: A New Generation (2021)  459/83   
200     Sunny Starscout  My Little Pony: A New Generation (2021)  459/83   
1299             Mothra                             MonsterVerse    73/5   
1306             Mothra                             MonsterVerse    73/5   
1499        Nyota Uhura              Star Trek (Kelvin Timeline)    63/3   
...                 ...                                      ...     ...   
167777          Asenath                   Joseph: King of Dreams     7/0   
167799     Kengo Kosaka                  Parasite in Love (2019)     7/0   
167810         Frédéric      Eric Rohmer's Love In The Afternoon     7/0   
167869     Meera Thapar                  Jab Tak Hai Jaan (2012)     7/0   
167876          Vanessa                        By the Sea (2015)     7/0   

        MBTI Enneagram category  \
199     ENFJ       1w2    movie   
200     ENFJ       1w2    movie   
1299    ENFJ       2w1    movie   
1306    ENFJ       2w1    movie   
1499    ENFJ       2w3    movie   
...      ...       ...      ...   
167777  INFJ       XwX    movie   
167799  INFJ       9w8    movie   
167810  INFJ       5w4    movie   
167869  INFJ       2w3    movie   
167876  INFJ       4w3    movie   

                                                  img_src  Page  
199     https://static1.personality-database.com/profi...     4  
200     https://static1.personality-database.com/profi...     5  
1299    https://static1.personality-database.com/profi...    26  
1306    https://static1.personality-database.com/profi...    27  
1499    https://static1.personality-database.com/profi...    30  
...                                                   ...   ...  
167777  https://pdb-web-static.s3.us-east-2.amazonaws....   206  
167799  https://static1.personality-database.com/profi...   206  
167810  https://static1.personality-database.com/profi...   207  
167869  https://pdb-web-static.s3.us-east-2.amazonaws....   208  
167876  https://static1.personality-database.com/profi...   208  

[5596 rows x 8 columns]

#### df 전체에 카테고리 함수 적용하고, movie만 뽑아보자

In [152]:
df['category'] = df['category'].map(category_dic)

In [153]:
df.head()

Character                Contents      Votes  MBTI Enneagram  \
0              Makima            Chainsaw Man   16.1k/2k  ENFJ       2w1   
1  Daenerys Targaryen  Game of Thrones (2011)   5.9k/940  ENFJ       1w2   
2      Tanjiro Kamado        Kimetsu no Yaiba   5.7k/799  ENFJ       1w2   
3        Oikawa Tooru               Haikyuu!!   5.6k/683  ENFJ       3w4   
4          Homelander    The Boys (2019-2022)  5.2k/1.9k  ENFJ       3w4   

        category                                            img_src  Page  
0  anime & manga  https://static1.personality-database.com/profi...     1  
1     television  https://static1.personality-database.com/profi...     1  
2  anime & manga  https://static1.personality-database.com/profi...     1  
3  anime & manga  https://static1.personality-database.com/profi...     1  
4     television  https://static1.personality-database.com/profi...     1

In [154]:
movie_df = df[df['category'] == 'movie']

In [155]:
movie_df # 19833 rows × 8 columns

Character                                 Contents  \
14           Moana Waialiki                             Moana (2016)   
16                      Joy                        Inside Out (2015)   
21               Judy Hopps                          Zootopia (2016)   
33               Elle Woods                    Legally Blonde (2001)   
37               Neil Perry                Dead Poets Society (1989)   
...                     ...                                      ...   
167965        Berkeley Cole                     Out of Africa (1985)   
167972             The Girl  A Girl Walks Home Alone at Night (2014)   
167975                Urzah                  The Dark Crystal (1982)   
167982           Muhsin Bey                               Muhsin Bey   
167984  Dr. Leonora Orantes                         Contagion (2011)   

           Votes  MBTI Enneagram category  \
14      2.7k/477  ENFJ       7w6    movie   
16      2.5k/358  ENFJ       7w6    movie   
21      2.3k/350  ENFJ       1w2    movie   
33      1.9k/227  ENFJ       3w2    movie   
37      1.7k/222  ENFJ       4w3    movie   
...          ...   ...       ...      ...   
167965       6/0  INFJ       1w9    movie   
167972       6/0  INFJ       4w5    movie   
167975       6/0  INFJ       9w1    movie   
167982       6/0  INFJ       1w2    movie   
167984       6/0  INFJ       9w1    movie   

                                                  img_src  Page  
14      https://static1.personality-database.com/profi...     1  
16      https://static1.personality-database.com/profi...     1  
21      https://static1.personality-database.com/profi...     1  
33      https://static1.personality-database.com/profi...     1  
37      https://static1.personality-database.com/profi...     1  
...                                                   ...   ...  
167965  https://static1.personality-database.com/profi...   210  
167972  https://static1.personality-database.com/profi...   210  
167975  https://static1.personality-database.com/profi...   210  
167982  https://static1.personality-database.com/profi...   210  
167984  https://static1.personality-database.com/profi...   210  

[19833 rows x 8 columns]

=> movie 카테고리로 줄이니 mbti, 애니어그램 데이터가 얼마 없다...

### 1-2. 데이터 EDA

Page 제거하고 카테고리 movie만 담은 df로 다시 EDA

In [156]:
import copy
df3 = copy.deepcopy(df)

In [157]:
df3.drop(columns='Page', inplace=True)

In [158]:
df3.head()

Character                Contents      Votes  MBTI Enneagram  \
0              Makima            Chainsaw Man   16.1k/2k  ENFJ       2w1   
1  Daenerys Targaryen  Game of Thrones (2011)   5.9k/940  ENFJ       1w2   
2      Tanjiro Kamado        Kimetsu no Yaiba   5.7k/799  ENFJ       1w2   
3        Oikawa Tooru               Haikyuu!!   5.6k/683  ENFJ       3w4   
4          Homelander    The Boys (2019-2022)  5.2k/1.9k  ENFJ       3w4   

        category                                            img_src  
0  anime & manga  https://static1.personality-database.com/profi...  
1     television  https://static1.personality-database.com/profi...  
2  anime & manga  https://static1.personality-database.com/profi...  
3  anime & manga  https://static1.personality-database.com/profi...  
4     television  https://static1.personality-database.com/profi...

In [159]:
movie_df3 = df3[df3['category'] == 'movie']

In [160]:
movie_df3

Character                                 Contents  \
14           Moana Waialiki                             Moana (2016)   
16                      Joy                        Inside Out (2015)   
21               Judy Hopps                          Zootopia (2016)   
33               Elle Woods                    Legally Blonde (2001)   
37               Neil Perry                Dead Poets Society (1989)   
...                     ...                                      ...   
167965        Berkeley Cole                     Out of Africa (1985)   
167972             The Girl  A Girl Walks Home Alone at Night (2014)   
167975                Urzah                  The Dark Crystal (1982)   
167982           Muhsin Bey                               Muhsin Bey   
167984  Dr. Leonora Orantes                         Contagion (2011)   

           Votes  MBTI Enneagram category  \
14      2.7k/477  ENFJ       7w6    movie   
16      2.5k/358  ENFJ       7w6    movie   
21      2.3k/350  ENFJ       1w2    movie   
33      1.9k/227  ENFJ       3w2    movie   
37      1.7k/222  ENFJ       4w3    movie   
...          ...   ...       ...      ...   
167965       6/0  INFJ       1w9    movie   
167972       6/0  INFJ       4w5    movie   
167975       6/0  INFJ       9w1    movie   
167982       6/0  INFJ       1w2    movie   
167984       6/0  INFJ       9w1    movie   

                                                  img_src  
14      https://static1.personality-database.com/profi...  
16      https://static1.personality-database.com/profi...  
21      https://static1.personality-database.com/profi...  
33      https://static1.personality-database.com/profi...  
37      https://static1.personality-database.com/profi...  
...                                                   ...  
167965  https://static1.personality-database.com/profi...  
167972  https://static1.personality-database.com/profi...  
167975  https://static1.personality-database.com/profi...  
167982  https://static1.personality-database.com/profi...  
167984  https://static1.personality-database.com/profi...  

[19833 rows x 7 columns]

In [161]:
name_list = movie_df3['Character'].value_counts()
name_list[name_list > 1] # 총 3251개의 name이 겹친다.

Tom                          24
Paul                         20
Alex                         18
Sam                          17
Lucy                         16
                             ..
Peter Parker "Spider-Man"     2
Jason                         2
Big Floppa                    2
Control                       2
Ghost / Ronnie                2
Name: Character, Length: 3251, dtype: int64

In [170]:
movie_df3[movie_df3.duplicated(['Character', 'Contents'], keep=False)] # Character 컬럼과 Contents 컬럼 모두가 겹치는 데이터는 총 5558행

Character                                 Contents   Votes  \
199     Sunny Starscout  My Little Pony: A New Generation (2021)  459/83   
200     Sunny Starscout  My Little Pony: A New Generation (2021)  459/83   
1299             Mothra                             MonsterVerse    73/5   
1306             Mothra                             MonsterVerse    73/5   
1499        Nyota Uhura              Star Trek (Kelvin Timeline)    63/3   
...                 ...                                      ...     ...   
167777          Asenath                   Joseph: King of Dreams     7/0   
167799     Kengo Kosaka                  Parasite in Love (2019)     7/0   
167810         Frédéric      Eric Rohmer's Love In The Afternoon     7/0   
167869     Meera Thapar                  Jab Tak Hai Jaan (2012)     7/0   
167876          Vanessa                        By the Sea (2015)     7/0   

        MBTI Enneagram category  \
199     ENFJ       1w2    movie   
200     ENFJ       1w2    movie   
1299    ENFJ       2w1    movie   
1306    ENFJ       2w1    movie   
1499    ENFJ       2w3    movie   
...      ...       ...      ...   
167777  INFJ       XwX    movie   
167799  INFJ       9w8    movie   
167810  INFJ       5w4    movie   
167869  INFJ       2w3    movie   
167876  INFJ       4w3    movie   

                                                  img_src  
199     https://static1.personality-database.com/profi...  
200     https://static1.personality-database.com/profi...  
1299    https://static1.personality-database.com/profi...  
1306    https://static1.personality-database.com/profi...  
1499    https://static1.personality-database.com/profi...  
...                                                   ...  
167777  https://pdb-web-static.s3.us-east-2.amazonaws....  
167799  https://static1.personality-database.com/profi...  
167810  https://static1.personality-database.com/profi...  
167869  https://pdb-web-static.s3.us-east-2.amazonaws....  
167876  https://static1.personality-database.com/profi...  

[5558 rows x 7 columns]

#### 왜 중복했는지 확인해보자. 단순하게 페이지만 달라서 중복됐던 건지?

In [171]:
movie_df3[movie_df3.duplicated(['Character', 'Contents'], keep=False)].sort_values(by='Character')

Character                                      Contents  \
59541    "Arthur Slugworth"  Willy Wonka and the Chocolate Factory (1971)   
59367    "Arthur Slugworth"  Willy Wonka and the Chocolate Factory (1971)   
17568                "Iris"                       The Midnight Sky (2020)   
17411                "Iris"                       The Midnight Sky (2020)   
140191  "Joliet" Jake Blues                     The Blues Brothers (1980)   
...                     ...                                           ...   
116352         भालू (Baloo)                        The Jungle Book (1967)   
6340           “Grandpappy”                   A House on the Bayou (2021)   
6356           “Grandpappy”                   A House on the Bayou (2021)   
135349     “Herm” Stockburn                              The Ridiculous 6   
135526     “Herm” Stockburn                              The Ridiculous 6   

        Votes  MBTI Enneagram category  \
59541    12/3  INTJ       4w5    movie   
59367    12/3  INTJ       4w5    movie   
17568    15/1  INFP       4w5    movie   
17411    15/1  INFP       4w5    movie   
140191   33/0  ESTP       8w7    movie   
...       ...   ...       ...      ...   
116352  162/6  ESFP       9w8    movie   
6340     11/0  XXXX       9w8    movie   
6356     11/0  XXXX       9w8    movie   
135349    8/0  ISFP       9w8    movie   
135526    8/0  ISFP       9w8    movie   

                                                  img_src  
59541   https://static1.personality-database.com/profi...  
59367   https://static1.personality-database.com/profi...  
17568   https://static1.personality-database.com/profi...  
17411   https://static1.personality-database.com/profi...  
140191  https://static1.personality-database.com/profi...  
...                                                   ...  
116352  https://static1.personality-database.com/profi...  
6340    https://pdb-web-static.s3.us-east-2.amazonaws....  
6356    https://pdb-web-static.s3.us-east-2.amazonaws....  
135349  https://static1.personality-database.com/profi...  
135526  https://static1.personality-database.com/profi...  

[5558 rows x 7 columns]

In [181]:
movie_df3[movie_df3.duplicated(['Character', 'Contents'], keep=False)].value_counts()

Character                        Contents                                                     Votes  MBTI  Enneagram  category  img_src                                                                                     
Private Snowball                 Full Metal Jacket (1987)                                     8/0    ISTJ  8w7        movie     https://static1.personality-database.com/profile_images/e3e3331f26594b4ba5b3c077c09c9c97.png    3
Jim Prideaux                     Tinker Tailor Soldier Spy (2011)                             11/0   ISTP  5w4        movie     https://static1.personality-database.com/profile_images/c8b524bbddb6476ea588c07af863e076.png    3
Descendants 2 - It's Goin' Down  Descendants (Franchise)                                      16/0   ESTJ  8w7        movie     https://pdb-web-static.s3.us-east-2.amazonaws.com/avatars/profile_avatar.png                    3
Ludvig                           Flåklypa « The Pinchcliffe Grand Prix» and the other movies  7/0    

In [183]:
movie_df3[movie_df3.duplicated(['Character', 'Contents'])].sort_values(by='Character')

Character  \
59541                    "Arthur Slugworth"   
17568                                "Iris"   
140226                  "Joliet" Jake Blues   
80646             "Mr. Fabulous" Alan Rubin   
71415             "Summer Hartley"/ Natasha   
...                                     ...   
113478  Федор Прокопенко \ Fedor Prokopenko   
97456                          Яна Семакина   
116352                         भालू (Baloo)   
6356                           “Grandpappy”   
135526                     “Herm” Stockburn   

                                            Contents  Votes  MBTI Enneagram  \
59541   Willy Wonka and the Chocolate Factory (1971)   12/3  INTJ       4w5   
17568                        The Midnight Sky (2020)   15/1  INFP       4w5   
140226                     The Blues Brothers (1980)   33/0  ESTP       8w7   
80646                      The Blues Brothers (1980)   11/0  ISTJ       3w2   
71415                              Definitely, Maybe    8/0  ENTJ       3w4   
...                                              ...    ...   ...       ...   
113478         Grom: Boyhood (Гром: Трудное детство)    7/1  ESFJ       2w3   
97456                               Univer. New Dorm   31/0  ISFJ       1w2   
116352                        The Jungle Book (1967)  162/6  ESFP       9w8   
6356                     A House on the Bayou (2021)   11/0  XXXX       9w8   
135526                              The Ridiculous 6    8/0  ISFP       9w8   

       category                                            img_src  
59541     movie  https://static1.personality-database.com/profi...  
17568     movie  https://static1.personality-database.com/profi...  
140226    movie  https://static1.personality-database.com/profi...  
80646     movie  https://static1.personality-database.com/profi...  
71415     movie  https://static1.personality-database.com/profi...  
...         ...                                                ...  
113478    movie  https://static1.personality-database.com/profi...  
97456     movie  https://static1.personality-database.com/profi...  
116352    movie  https://static1.personality-database.com/profi...  
6356      movie  https://pdb-web-static.s3.us-east-2.amazonaws....  
135526    movie  https://static1.personality-database.com/profi...  

[2830 rows x 7 columns]

movie_df3은 카테고리가 movie인 데이터만 담은 데이터프레임

### 2. 데이터 클렌징

#### 단순하게 중복 제거 시

In [48]:
import copy
df2 = copy.deepcopy(df)

In [49]:
df2.drop(columns='Page', inplace=True)

In [50]:
df2.head()

Character                Contents      Votes  MBTI Enneagram  \
0              Makima            Chainsaw Man   16.1k/2k  ENFJ       2w1   
1  Daenerys Targaryen  Game of Thrones (2011)   5.9k/940  ENFJ       1w2   
2      Tanjiro Kamado        Kimetsu no Yaiba   5.7k/799  ENFJ       1w2   
3        Oikawa Tooru               Haikyuu!!   5.6k/683  ENFJ       3w4   
4          Homelander    The Boys (2019-2022)  5.2k/1.9k  ENFJ       3w4   

                                            category  \
0  https://pdb-web-static.s3.us-east-2.amazonaws....   
1  https://pdb-web-static.s3.us-east-2.amazonaws....   
2  https://pdb-web-static.s3.us-east-2.amazonaws....   
3  https://pdb-web-static.s3.us-east-2.amazonaws....   
4  https://pdb-web-static.s3.us-east-2.amazonaws....   

                                             img_src  
0  https://static1.personality-database.com/profi...  
1  https://static1.personality-database.com/profi...  
2  https://static1.personality-database.com/profi...  
3  https://static1.personality-database.com/profi...  
4  https://static1.personality-database.com/profi...

In [51]:
df2_drop_dupli = df2.drop_duplicates(subset=['Character', 'Contents'])

In [52]:
df2_drop_dupli.head()

Character                Contents      Votes  MBTI Enneagram  \
0              Makima            Chainsaw Man   16.1k/2k  ENFJ       2w1   
1  Daenerys Targaryen  Game of Thrones (2011)   5.9k/940  ENFJ       1w2   
2      Tanjiro Kamado        Kimetsu no Yaiba   5.7k/799  ENFJ       1w2   
3        Oikawa Tooru               Haikyuu!!   5.6k/683  ENFJ       3w4   
4          Homelander    The Boys (2019-2022)  5.2k/1.9k  ENFJ       3w4   

                                            category  \
0  https://pdb-web-static.s3.us-east-2.amazonaws....   
1  https://pdb-web-static.s3.us-east-2.amazonaws....   
2  https://pdb-web-static.s3.us-east-2.amazonaws....   
3  https://pdb-web-static.s3.us-east-2.amazonaws....   
4  https://pdb-web-static.s3.us-east-2.amazonaws....   

                                             img_src  
0  https://static1.personality-database.com/profi...  
1  https://static1.personality-database.com/profi...  
2  https://static1.personality-database.com/profi...  
3  https://static1.personality-database.com/profi...  
4  https://static1.personality-database.com/profi...

In [56]:
df2_drop_dupli[df2_drop_dupli.duplicated(subset=['Character','Contents'])]

Empty DataFrame
Columns: [Character, Contents, Votes, MBTI, Enneagram, category, img_src]
Index: []

In [57]:
print(df.shape)
print(df2_drop_dupli.shape)

(168000, 8)
(143333, 7)


## 카테고리 movie인 데이터만 우선 제목 맞춰주기

In [4]:
df

Character                  Contents  \
0                                          Makima              Chainsaw Man   
1                              Daenerys Targaryen    Game of Thrones (2011)   
2                                  Tanjiro Kamado          Kimetsu no Yaiba   
3                                    Oikawa Tooru                 Haikyuu!!   
4                                      Homelander      The Boys (2019-2022)   
...                                           ...                       ...   
167995                            Amelia Scanwell                   Harlots   
167996                                        Jim              Toaster Dude   
167997                                  Channeler      Pokémon Red and Blue   
167998                                     Raksha           The Jungle Book   
167999  Abdul Qadir Al Jilani, Hanbali Theologian  Islamic Scholars (Sunni)   

            Votes  MBTI Enneagram  \
0        16.1k/2k  ENFJ       2w1   
1        5.9k/940  ENFJ       1w2   
2        5.7k/799  ENFJ       1w2   
3        5.6k/683  ENFJ       3w4   
4       5.2k/1.9k  ENFJ       3w4   
...           ...   ...       ...   
167995        6/0  INFJ       9w1   
167996        6/0  INFJ       5w4   
167997        6/2  INFJ       2w1   
167998        6/0  INFJ       XwX   
167999        6/0  INFJ       9w1   

                                                 category  \
0       https://pdb-web-static.s3.us-east-2.amazonaws....   
1       https://pdb-web-static.s3.us-east-2.amazonaws....   
2       https://pdb-web-static.s3.us-east-2.amazonaws....   
3       https://pdb-web-static.s3.us-east-2.amazonaws....   
4       https://pdb-web-static.s3.us-east-2.amazonaws....   
...                                                   ...   
167995  https://pdb-web-static.s3.us-east-2.amazonaws....   
167996  https://pdb-web-static.s3.us-east-2.amazonaws....   
167997  https://pdb-web-static.s3.us-east-2.amazonaws....   
167998  https://pdb-web-static.s3.us-east-2.amazonaws....   
167999  https://pdb-web-static.s3.us-east-2.amazonaws....   

                                                  img_src  Page  
0       https://static1.personality-database.com/profi...     1  
1       https://static1.personality-database.com/profi...     1  
2       https://static1.personality-database.com/profi...     1  
3       https://static1.personality-database.com/profi...     1  
4       https://static1.personality-database.com/profi...     1  
...                                                   ...   ...  
167995  https://pdb-web-static.s3.us-east-2.amazonaws....   210  
167996  https://static1.personality-database.com/profi...   210  
167997  https://static1.personality-database.com/profi...   210  
167998  https://static1.personality-database.com/profi...   210  
167999  https://static1.personality-database.com/profi...   210  

[168000 rows x 8 columns]

In [10]:
df.shape

(168000, 8)

In [19]:
df['category'] = df['category'].map(category_dic)

In [20]:
import copy
movie = copy.deepcopy(df)

In [21]:
movie = movie.drop(columns='Page')

In [22]:
movie = movie.drop_duplicates()

In [23]:
movie.shape

(143470, 7)

In [24]:
movie = movie[movie['category'] == 'movie']

In [25]:
movie.shape

(17007, 7)

In [29]:
ml_df = pd.read_csv("/Users/jeonhaelee/Documents/boostcamp/project/code/data/ml-25m/movies.csv")
ml_df.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [30]:
ml_title = list(ml_df['title'].unique())
movie_title = list(movie['Contents'].unique())
print(len(ml_title), len(movie_title))

62325 4291


In [32]:
check_title = []
for title in movie_title:
    if title not in ml_title:
        check_title.append(title)
        print(title, "isn't in movielens title ‼️")
print("length :", len(check_title)) # 2619

Turning Red (2022) isn't in movielens title ‼️
Star Wars isn't in movielens title ‼️
Frozen (Franchise) isn't in movielens title ‼️
The Lion King (Franchise) isn't in movielens title ‼️
Avatar (Franchise) isn't in movielens title ‼️
The Hunger Games (Franchise) isn't in movielens title ‼️
Toy Story (Franchise) isn't in movielens title ‼️
Harry Potter / Wizarding World Films isn't in movielens title ‼️
Emma. (2020) isn't in movielens title ‼️
The Matrix Trilogy isn't in movielens title ‼️
Descendants (Franchise) isn't in movielens title ‼️
My Little Pony: A New Generation (2021) isn't in movielens title ‼️
Fantastic Beasts isn't in movielens title ‼️
To All the Boys (trilogy) isn't in movielens title ‼️
The Mitchells vs. the Machines (2021) isn't in movielens title ‼️
Dune (2021) isn't in movielens title ‼️
Encanto (2021) isn't in movielens title ‼️
The Black Phone isn't in movielens title ‼️
Raya and the Last Dragon (2021) isn't in movielens title ‼️
Do Revenge (2022) isn't in movielen

#### 제목을 다 소문자로 바꿔서 비교해주자

In [36]:
ml_title2 = [title.lower() for title in ml_title]
movie_title2 = [title.lower() for title in movie_title]
check_title_lower = []
for title in movie_title2:
    if title not in ml_title2:
        check_title_lower.append(title)
        print(title, "isn't in movielens title ‼️")
print("length :", len(check_title_lower)) # 2572

turning red (2022) isn't in movielens title ‼️
star wars isn't in movielens title ‼️
frozen (franchise) isn't in movielens title ‼️
the lion king (franchise) isn't in movielens title ‼️
avatar (franchise) isn't in movielens title ‼️
the hunger games (franchise) isn't in movielens title ‼️
toy story (franchise) isn't in movielens title ‼️
harry potter / wizarding world films isn't in movielens title ‼️
emma. (2020) isn't in movielens title ‼️
the matrix trilogy isn't in movielens title ‼️
descendants (franchise) isn't in movielens title ‼️
my little pony: a new generation (2021) isn't in movielens title ‼️
fantastic beasts isn't in movielens title ‼️
to all the boys (trilogy) isn't in movielens title ‼️
the mitchells vs. the machines (2021) isn't in movielens title ‼️
dune (2021) isn't in movielens title ‼️
encanto (2021) isn't in movielens title ‼️
the black phone isn't in movielens title ‼️
raya and the last dragon (2021) isn't in movielens title ‼️
do revenge (2022) isn't in movielen

#### 제목에 있는 연도를 없애보자

In [43]:
def del_year(title):
    try:
        idx = title.index("(")
        return title[:idx].strip()
    except:
        return title.strip()

In [44]:
ml_title3 = list(map(del_year, ml_title2))
movie_title3 = list(map(del_year, movie_title2))
check_title_lower2 = []
for title in movie_title3:
    if title not in ml_title3:
        check_title_lower2.append(title)
        print(title, "isn't in movielens title ‼️")
print("length :", len(check_title_lower2)) # 1554

turning red isn't in movielens title ‼️
star wars isn't in movielens title ‼️
harry potter / wizarding world films isn't in movielens title ‼️
emma. isn't in movielens title ‼️
the matrix trilogy isn't in movielens title ‼️
my little pony: a new generation isn't in movielens title ‼️
fantastic beasts isn't in movielens title ‼️
to all the boys isn't in movielens title ‼️
the mitchells vs. the machines isn't in movielens title ‼️
encanto isn't in movielens title ‼️
the black phone isn't in movielens title ‼️
raya and the last dragon isn't in movielens title ‼️
do revenge isn't in movielens title ‼️
the chronicles of narnia isn't in movielens title ‼️
a streetcar named desire isn't in movielens title ‼️
the adventures of sharkboy and lavagirl isn't in movielens title ‼️
the bad guys isn't in movielens title ‼️
don't worry darling isn't in movielens title ‼️
the hobbit isn't in movielens title ‼️
soul isn't in movielens title ‼️
jurassic park / jurassic world isn't in movielens title ‼️
n

#### 무비렌즈 제목이랑 유사도 측정해서 비슷한 제목 뽑아보자
ml_title3, movie_title3

In [45]:
def content_based_filtering_jaccard(title_list, title):
    topn=3
    target_split_set = set(title.split())
    sim_list = []
    sim_df = pd.DataFrame(title_list, columns=['title'])
    for idx, movie in enumerate(title_list):
        title_split_set = set(movie.split())
        title_intersection = target_split_set.intersection(title_split_set)
        jac_sim = float(len(title_intersection)) / (len(target_split_set) + len(title_split_set) - len(title_intersection))
        sim_list.append(jac_sim)
    sim_df['jaccard_similarity'] = sim_list

    return sim_df.sort_values('jaccard_similarity', ascending=False).reset_index(drop=True)[:topn]

In [50]:
title_list = ml_title3
title = check_title_lower2[0] # 'turning red'
print('crawling title :', title)
content_based_filtering_jaccard(title_list, title)

crawling title : turning red


title  jaccard_similarity
0          red            0.500000
1          red            0.500000
2  red sparrow            0.333333

In [53]:
title_list = ml_title3
title = check_title_lower2[1] # 'star wars'
print('crawling title :', title)
content_based_filtering_jaccard(title_list, title)

crawling title : star wars


title  jaccard_similarity
0       star wars downunder            0.666667
1                      star            0.500000
2  robot chicken: star wars            0.500000

In [54]:
title_list = ml_title3
title = check_title_lower2[2] # 'harry potter'
print('crawling title :', title)
content_based_filtering_jaccard(title_list, title)

crawling title : harry potter / wizarding world films


title  jaccard_similarity
0   harry potter and the half-blood prince            0.200000
1    harry potter and the sorcerer's stone            0.200000
2  harry potter and the chamber of secrets            0.181818

In [55]:
title_list = ml_title3
title = check_title_lower2[3] # 'emma.'
print('crawling title :', title)
content_based_filtering_jaccard(title_list, title)

crawling title : emma.


title  jaccard_similarity
0                     toy story                 0.0
1  frankenstein: the true story                 0.0
2               between friends                 0.0